django/urls/base.py

```diff
@@ -179,3 +183,50 @@ def translate_url(url, lang_code):
             else:
                 url = urlunsplit((parsed.scheme, parsed.netloc, url, parsed.query, parsed.fragment))
     return url
+
+
+_PATH_PARAMETER_COMPONENT_RE = re.compile(
+    '<(?:(?P<converter>[^:]+):)?(?P<parameter>[A-Za-z0-9_]+)>'
+)
+
+
+def _route_to_regex(route):
+    parts = ['^']
+    converters = {}
+    while True:
```

Comment:
Not sure if it actually would be any more elegant, but it's at least worth an investigation. It might also be a bit more efficient as well.